enum_methods.rb

```diff
@@ -0,0 +1,155 @@
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    each do |i|
+      yield(i)
+    end
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    my_each do |element|
+      yield(element, i)
+      i += 1
+    end
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    new_arr = []
+    my_each do |i|
+      new_arr << i if yield(i)
+    end
+    new_arr
+  end
+
+  def my_all?(arg = nil)
+    unless arg.nil?
+      if arg.is_a?(Regexp)
+        my_each { |ind| return false unless ind.match(arg) }
+      elsif arg.is_a?(Module)
+        my_each { |ind| return false unless ind.is_a?(arg) }
+      else
+        my_each { |ind| return false if ind != arg }
+      end
+      return true
+    end
+
+    unless block_given?
+
+      my_each { |ind| return false if ind == false || ind.nil? }
+      return true
+    end
+
+    my_each do |i|
+      t_f = yield(i)
+      return false unless t_f
+    end
+    true
+  end
```

Comment:
**Optional** Your method is too complex. you can achieve the same result if you think a little more and refactor your code to make it simpler.